File: intern-frontend/src/app/spending-list/service/spending.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Spending} from '../model/spending.model';

@Injectable({
  providedIn: 'root',
})
export class SpendingService {
  private spendingUrl: string = '/api/v1/spendings';

  constructor(private http: HttpClient) {
  }

  getAllSpendings(): Observable<Spending[]> {
    return this.http.get<Spending[]>(`${this.spendingUrl}`);
  }

  getAllSpendingsByEmployeeId(employeeId: number): Observable<Spending[]> {
    return this.http.get<Spending[]>(`${this.spendingUrl}/employee/${employeeId}`);
  }

  createSpending(data: Spending): Observable<Spending> {
    return this.http.post<Spending>(`${this.spendingUrl}`, data);
  }


  updateSpending(data: Spending): Observable<Spending> {
    return this.http.put<Spending>(`${this.spendingUrl}/${data.id}`, data);
  }


  deleteSpending(id: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.spendingUrl}/${id}`);
  }

  changeSpendingStatus(id: number): Observable<Spending> {
    return this.http.get<Spending>(`${this.spendingUrl}/changeSpendingStatus/${id}`);
  }

  getPendingSpendings(): Observable<Spending[]> {
    return this.http.get<Spending[]>(`${this.spendingUrl}/pendingSpendings`);
  }

  countPendingSpendings(): Observable<number> {
    return this.http.get<number>(`${this.spendingUrl}/countPendingSpendings`);
  }

  getTodayAndApprovedSpendings(): Observable<Spending[]> {
    return this.http.get<Spending[]>(`${this.spendingUrl}/todayAndApprovedSpendings`);
  }

  countTodayAndApprovedSpendings(): Observable<number> {
    return this.http.get<number>(`${this.spendingUrl}/countTodayAndApprovedSpendings`);
  }

  countAllSpendingsByEmployeeId(employeeId: number): Observable<number> {
    return this.http.get<number>(`${this.spendingUrl}/count/${employeeId}`);
  }

  countAllSpendings(): Observable<number> {
    return this.http.get<number>(`${this.spendingUrl}/count`);
  }
}

File: intern-backend/src/main/java/com/internBackend/spending/service/SpendingService.java
package com.internBackend.spending.service;
import com.internBackend.spending.entity.Spending;

import java.util.List;

public interface SpendingService {
    List<Spending> fetchAllSpendings();
    List<Spending> getAllSpendingsByEmployeeId(Long employeeId);
    Long countAllSpendings();
    Long countAllApprovedSpendings();
    Long countAllApprovedSpendingsByEmployeeId(Long employeeId);
    Long countUsedSpendingsByEmployeeId(Long employeeId);


    Integer getRemainingSpendingByEmployeeId(Long employeeId);

    Spending fetchSpendingById(Long id);
    Spending createSpending(Spending spending);
    Spending updateSpending(Spending spending);
    String deleteSpending(Spending spending);
    Spending changeSpendingStatus(Long id);

    List<Spending> getPendingSpendings();
    Long countPendingSpendings();
    List<Spending> getTodayAndApprovedSpendings();
    Long countTodayAndApprovedSpendings();

    Long countAllSpendingsByEmployeeId(Long employeeId);
}
